### PR TITLE
{Identity} `az identity`: Upgrade api-version 2023-01-31 for managed identity

### PR DESCRIPTION
**Pull request by yanzhudd**
**Related command**

`az identity`

**Description**

MSI version 2023-01-31 is released. 
Close: #25380

**Testing Guide**

**History Notes**

{Identity} `az identity`: Upgrade api-version 2023-01-31 of managed identity for commands except `az identity list-resources`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Identity

File: src/azure-cli/azure/cli/command_modules/identity/commands.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def load_command_table(self, _):
         g.show_command('show', 'get')
         g.command('delete', 'delete')
         g.custom_command('list', 'list_user_assigned_identities')
-        g.command('list-resources', 'list_associated_resources', min_api='2021-09-30-preview')
+        g.custom_command('list-resources', 'list_identity_resources', min_api='2021-09-30-preview')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/identity/commands.py:35]:
```suggestion
        g.custom_command('list-resources', 'list_identity_resources', min_api='2021-09-30-preview', deprecate_info=c.deprecate(hide=True))
```

**Comment by jiasli** [comment]
The name "MSI" has been deprecated long ago. We should consider renaming it to "managed identity":

https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview

> Managed identities for Azure resources is the new name for the service formerly known as Managed Service Identity (MSI).

File: src/azure-cli/azure/cli/command_modules/identity/_client_factory.py [hunk 2]
```diff
@@ -4,10 +4,14 @@
...
-def _msi_client_factory(cli_ctx, **_):
+def _msi_client_factory(cli_ctx, api_version=None, **_):
     from azure.cli.core.profiles import ResourceType
     from azure.cli.core.commands.client_factory import get_mgmt_service_client
-    return get_mgmt_service_client(cli_ctx, ResourceType.MGMT_MSI)
+    return get_mgmt_service_client(cli_ctx, ResourceType.MGMT_MSI, api_version=api_version)
+
+
+def _msi_list_resources_client(cli_ctx, **_):
+    return _msi_client_factory(cli_ctx, api_version='2021-09-30-preview').user_assigned_identities
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/identity/_client_factory.py:14]: Please add a comment to explain why we need to specify the ad-hoc api-version

Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/identity/_client_factory.py:14]: Please use the latest available api-version
```suggestion
    return _msi_client_factory(cli_ctx, api_version='2022-01-31-preview').user_assigned_identities
```

Comment (yanzhudd) [reply src/azure-cli/azure/cli/command_modules/identity/_client_factory.py:14]: Hi @isolenov,
we CLI use the api-version `2022-01-31-preview` before upgrading and it supports `listAssociatedResources` command. Could you please help to confirm which version (`2022-01-31-preview` or `2021-09-30-preview`) should be used for `listAssociatedResources` command?

Comment (isolenov) [reply src/azure-cli/azure/cli/command_modules/identity/_client_factory.py:14]: @yanzhudd @zhoxing-ms please use latest preview version `2022-01-31-preview` for `listAssociatedResources`

**Review by zhoxing-ms — APPROVED** [review]

**Review by necusjz — DISMISSED** [review]

**Review by evelyn-ys — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/identity/_client_factory.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/identity/_client_factory.py]: Ad-hoc API version can leads to `scripts/trim_sdk.py` script failure.

It should be declared at

https://github.com/Azure/azure-cli/blob/085daccc19445387586e4f532b325f51e526eaff/src/azure-cli-core/azure/cli/core/profiles/_shared.py#L411-L428

**Review by jiasli — CHANGES_REQUESTED** [review]
Ad-hoc API version should be forbidden.

**Comment by yanzhudd** [comment]
> Ad-hoc API version should be forbidden unless declared.

@jiasli Since there is a breaking change in the new version of MSI (a command is not supported), ad-hoc api version here is used for avoiding the breaking change. And we have discussed with the service team, and it seems this approach work the best. 

The declaration for the ad-hoc api version is added, could you please help to review this change?

File: src/azure-cli/azure/cli/command_modules/identity/_client_factory.py (hunk 2 above)
Comment (yanzhudd) [reply src/azure-cli/azure/cli/command_modules/identity/_client_factory.py]: Hi @jiasli, declaration has been added here. Could you please review this change?

**Review by jiasli — APPROVED** [review]

**Review by necusjz — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py [hunk 3]
```diff
@@ -424,6 +424,9 @@ def default_api_version(self):
     ResourceType.MGMT_CONTAINERSERVICE: {
         # src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_custom.py:50
         'ManagedClusterAddonProfile': '2020-03-01',
+    },
+    ResourceType.MGMT_MSI: {
+        'user_assigned_identities': '2022-01-31-preview',
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/profiles/_shared.py:429]: It's recommended to add a comment showing where `2022-01-31-preview` is used, like L425.

In this case, it'll be

```py
        # src/azure-cli/azure/cli/command_modules/identity/_client_factory.py:18
```

**Pull request merged**